src/ovs/aca_ovs_l3_programmer.cpp

```diff
@@ -0,0 +1,262 @@
+    auto found_subnet = router_it->second.find(subnet_id);
+
+    if (found_subnet == router_it->second.end()) {
+      // subnet not found in this router, go look at the next router
+      continue;
+    } else {
+      // destination subnet found!
+      found_subnet_in_router = true;
+
+      // for each other subnet connected to this router, create the routing rule
+      for (auto subnet_it = router_it->second.begin();
+           subnet_it != router_it->second.end(); subnet_it++) {
+        // skip the destination neighbor subnet
+        if (subnet_it->first == subnet_id) {
+          continue;
+        }
+        ACA_LOG_DEBUG("subnet_id:%s\n ", subnet_it->first.c_str());
+
+        source_vlan_id = ACA_Vlan_Manager::get_instance().get_or_create_vlan_id(
+                subnet_it->second.vpc_id);
+
+        // essential rule to restore from neighbor host DVR mac to destination GW mac:
+        // not needed if the neighbor port is at the same compute host
+        if (!is_port_on_same_host) {
+          cmd_string = "add-flow br-int \"table=0,priority=25,dl_vlan=" +
+                       to_string(source_vlan_id) + ",dl_src=" + neighbor_host_dvr_mac +
+                       " actions=mod_dl_src:" + subnet_it->second.gateway_mac +
+                       " output:NORMAL\"";
+
+          ACA_OVS_L2_Programmer::get_instance().execute_openflow_command(
+                  cmd_string, culminative_time, overall_rc);
+        }
+
+        destination_vlan_id =
+                ACA_Vlan_Manager::get_instance().get_or_create_vlan_id(vpc_id);
+
+        // for the first implementation, we will go ahead and program the on demand routing rule here
+        // in the future, the programming of the on demand rule will be triggered by the first packet
+        // sent to openflow controller, that's ACA
+
+        // the openflow rule depends on whether the hosting ip is on this compute host or not
+        if (is_port_on_same_host) {
+          cmd_string = "add-flow br-tun \"table=0,priority=50,ip,dl_vlan=" +
+                       to_string(source_vlan_id) + ",nw_dst=" + virtual_ip +
+                       ",dl_dst=" + subnet_it->second.gateway_mac +
+                       " actions=mod_vlan_vid:" + to_string(destination_vlan_id) +
+                       ",mod_dl_dst:" + virtual_mac + ",output:IN_PORT\"";
+        } else {
+          cmd_string = "add-flow br-tun \"table=0,priority=50,ip,dl_vlan=" +
+                       to_string(source_vlan_id) + ",nw_dst=" + virtual_ip +
+                       ",dl_dst=" + subnet_it->second.gateway_mac +
+                       " actions=mod_vlan_vid:" + to_string(destination_vlan_id) +
+                       ",mod_dl_src:" + _host_dvr_mac +
+                       ",mod_dl_dst:" + virtual_mac + ",resubmit(,2)\"";
+        }
+
+        ACA_OVS_L2_Programmer::get_instance().execute_openflow_command(
+                cmd_string, culminative_time, overall_rc);
+      }
+      break;
+    }
+  }
```

Comment:
Why you put break in the end of for-loop?

Reply:
that's a great question, in this else {  // destination subnet found! ... break; } block, we found our interested router from _routers_table which has the destination subnet GW connected to it. Since each subnet GW can only be connected to one router, therefore, there is no point to look at other routers on the higher level for (auto router_it = _routers_table.begin(); loop, hence the "break;" to break it.